src/azure-cli-core/azure/cli/core/auth/binary_cache.py

```diff
@@ -0,0 +1,79 @@
+import collections.abc as collections
+import pickle
+
+from azure.cli.core.decorators import retry
+from knack.log import get_logger
+
+logger = get_logger(__name__)
+
+
+class BinaryCache(collections.MutableMapping):
+    """
+    Derived from azure.cli.core._session.Session.
+    A simple dict-like class that is backed by a binary file.
+
+    All direct modifications with `__setitem__` and `__delitem__` will save the file.
+    Indirect modifications should be followed by a call to `save`.
+    """
+
+    def __init__(self, file_name):
+        super().__init__()
+        self.filename = file_name
+        self.data = {}
+        self.load()
+
+    @retry()
+    def _load(self):
+        """Load cache with retry. If it still fails at last, raise the original exception as-is."""
+        try:
+            with open(self.filename, 'rb') as f:
+                return pickle.load(f)
+        except FileNotFoundError:
+            # The cache file has not been created. This is expected. No need to retry.
+            logger.debug("%s not found. Using a fresh one.", self.filename)
+            return {}
+
+    def load(self):
+        logger.debug("load: %s", self.filename)
+        try:
+            self.data = self._load()
+        except (pickle.UnpicklingError, EOFError) as ex:
+            # We still get exception after retry:
+            # - pickle.UnpicklingError is caused by corrupted cache file, perhaps due to concurrent writes.
+            # - EOFError is caused by empty cache file created by other az instance, but hasn't been filled yet.
+            logger.debug("Failed to load cache: %s. Using a fresh one.", ex)
+            self.data = {}  # Ignore a non-existing or corrupted http_cache
+
+    @retry()
+    def _save(self):
```

Comment:
My gut feel tells me to retry the `_save` method, though I don't know what exception will be raised here.

Reply:
Wrote an example to constantly simulate concurrent writes which can corrupt a file:

```py
# proc.py

import time
import sys

string_to_write = sys.argv[1]
with open(r"testfile.txt", 'w') as f:
    print(f"{string_to_write}: Sleeping for 5s before writing")
    time.sleep(5)
    f.write(string_to_write)

print(f"{string_to_write}: Finished writing")
```

```py
# caller.py

import subprocess
import time

proc1 = subprocess.Popen(['python', 'proc.py', 'abcdefghijklmnopqrstuvwxyz'])
time.sleep(2)
proc2 = subprocess.Popen(['python', 'proc.py', '0123456789'])

proc1.wait()
proc2.wait()

with open(r"testfile.txt") as f:
    print("Final result:", f.read())
```

Strings `abcdefghijklmnopqrstuvwxyz` and `0123456789` will be written to `testfile.txt` in the order of

```
|---|---|---|---|---|         abcdefghijklmnopqrstuvwxyz
        |---|---|---|---|---| 0123456789
```

Output:

```
python caller.py
abcdefghijklmnopqrstuvwxyz: Sleeping for 5s before writing
0123456789: Sleeping for 5s before writing
abcdefghijklmnopqrstuvwxyz: Finished writing
0123456789: Finished writing
Final result: 0123456789klmnopqrstuvwxyz
```